lib/wibox/widget/separator.lua

```diff
@@ -0,0 +1,207 @@
+-- @see thickness
+
+--- The separator border color.
+-- @beautiful beautiful.separator_border_color
+-- @param gears.color
+-- @see border_color
+
+--- The separator border width.
+-- @beautiful beautiful.separator_border_width
+-- @param number
+-- @see border_width
+
+--- The relative percentage covered by the bar.
+-- @beautiful beautiful.separator_span_ratio
+-- @tparam[opt=1] number A number between 0 and 1.
+
+--- The separator's color.
+-- @beautiful beautiful.separator_color
+-- @param string
+-- @see gears.color
+
+--- The separator's shape.
+--
+-- @beautiful beautiful.separator_shape
+-- @tparam function shape A valid shape function
+-- @see gears.shape
+
+local function draw_shape(self, _, cr, width, height, shape)
+    local bw = self._private.border_width or beautiful.separator_border_width or 0
+    local bc = self._private.border_color or beautiful.separator_border_color
+
+    cr:translate(bw/2, bw/2)
+
+    shape(cr, width-bw, height-bw)
+
+    if bw == 0 then
+        cr:fill()
+    elseif bc then
+        cr:fill_preserve()
+        cr:set_source(color(bc))
+        cr:set_line_width(bw)
+        cr:stroke()
+    end
+end
+
+local function draw_line(self, _, cr, width, height)
+    local thickness = self._private.thickness or beautiful.separator_thickness or 1
+
+    local orientation = self._private.orientation or (
+        width > height and "horizontal" or "vertical"
+    )
+
+    local span_ratio = self.span_ratio or 1
+
+    if orientation == "horizontal" then
+        local w = width*span_ratio
+        cr:rectangle((width-w)/2, height/2 - thickness/2, w, thickness)
+    else
+        local h = height*span_ratio
+        cr:rectangle(width/2 - thickness/2, (height-h)/2, thickness, h)
+    end
+
+    cr:fill()
+end
+
+local function draw(self, _, cr, width, height)
+    -- In case there is a specialized.
+    local draw_custom = self._private.draw or beautiful.separator_draw
+    if draw_custom then
+        return draw_custom(self, _, cr, width, height)
+    end
+
+    local col = self._private.color or beautiful.separator_color
+
+    if col then
+        cr:set_source(color(col))
+    end
+
+    local s = self._private.shape or beautiful.separator_shape
+
+    if s then
+        draw_shape(self, _, cr, width, height, s)
+    else
+        draw_line(self, _, cr, width, height)
+    end
+end
+
+local function fit(_, _, width, height)
+    return width, height
+end
+
+for _, prop in ipairs {"orientation", "color", "thickness", "span_ratio",
+                       "border_width", "border_color", "shape" } do
+    separator["set_"..prop] = function(self, value)
+        self._private[prop] = value
+        self:emit_signal("property::"..prop)
+        if prop == "data_list" then
+            self:emit_signal("property::data")
+        end
+        self:emit_signal("widget::redraw_needed")
+    end
+    separator["get_"..prop] = function(self)
+        return self._private[prop] or beautiful["separator_"..prop]
```

Comment:
Could/should all the code in the widget be changed to use this getter? So, for example, in line 163, instead of `self._private.shape or beautiful.separator_shape`, just use `self.shape`?
For example, right now the getter will "lie" about a separator's orientation if `beautiful.separator_orientation` is set, because the actual code using it doesn't check the beautiful value. Just using the getter everywhere would ensure nothing like this can happen.